### PR TITLE
Add reset op and accfg-insert-resets

### PR DESCRIPTION
**Pull request by AntonLydike**
This adds a reset op and a new pass called `accfg-insert-resets`, which adds an `accfg.reset` operation onto every dangling state.

Stacked onto #172

### CODE REVIEW

File: compiler/transforms/accfg_insert_resets.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                if not isinstance(val.type, val_type):
+                    continue
+                if val in seen:
+                    continue
+                seen.add(val)
+                fun(self, val, rewriter)
+
+        return match_and_rewrite
+
+    return wrapper
```
Comment (AntonLydike) [inline compiler/transforms/accfg_insert_resets.py:36]: I want to eventually move this into upstream xDSL

File: tests/filecheck/transforms/accfg-insert-resets.mlir
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:1]: Oh this one is in the other test as well?
Shouldn't  we also test this without overlap?

File: compiler/dialects/accfg.py [hunk 2]
```diff
@@ -430,12 +430,29 @@ def launch_field_items(self) -> Iterable[tuple[str, IntegerAttr]]:
...
+    in_state = operand_def(StateType)
+
+    assembly_format = "$in_state attr-dict `:` type($in_state)"
+
+    def __init__(self, in_state: Operation | SSAValue):
+        super().__init__(operands=[in_state])
+
+    def get_modified_fields(self) -> set[str]:
+        # TODO: implement this part
+        return set()
```
Comment (JosseVanDelm) [inline compiler/dialects/accfg.py:446]: why is this here?

File: compiler/transforms/accfg_insert_resets.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        inserts[block] = InsertCandidate(idx, use.operation)
+
+    # return all insertion candidates that we have
+    yield from (InsertPoint.after(cd.op) for cd in inserts.values())
+
+
+class InsertResetsForDanglingStatesPattern(RewritePattern):
+    """
+    This inspects all SSA values and inserts a `reset` operation if the value has no uses outside of `launch` and
+    `reset` operations.
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_insert_resets.py:97]: why also on reset operations? Wouldn't  this reset twice?

File: compiler/transforms/accfg_insert_resets.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    """
+
+    @ssa_val_rewrite_pattern(accfg.StateType)
+    def match_and_rewrite(self, val: SSAValue, rewriter: PatternRewriter, /):
+        # abort if:
+        #  - there are uses of the op
+        #  - and all of them are setups/reset ops
+        # this may not be correct, let's see...
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_insert_resets.py:105]: I'm not sure I understand this

**Review by JosseVanDelm — COMMENTED** [review]
I have a few questions

File: tests/filecheck/transforms/accfg-insert-resets.mlir
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:1]: Hmmm, probably yes. Will add!

File: tests/filecheck/transforms/accfg-insert-resets.mlir [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    %t = "accfg.launch"(%state) <{"param_names" = [], "accelerator" = "acc"}> : (!accfg.state<"acc">) -> !accfg.token<"acc">
+    "accfg.await"(%t) : (!accfg.token<"acc">) -> ()
+    return
+}
+
+// CHECK-LABEL: @with_uses
+// CHECK-NEXT: [[state:%\S+]] = accfg.setup "acc" to ("i" = %i : i32) : !accfg.state<"acc">
+// CHECK-NEXT: [[token:%\S+]] = "accfg.launch"(%state)
+// CHECK-NEXT: accfg.reset [[state]] : !accfg.state<"acc">
+// CHECK-NEXT: "accfg.await"([[token]])
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:49]: The `CHECK` lines don't need to match the whole line

File: compiler/dialects/accfg.py (hunk 2 above)
Comment (AntonLydike) [reply compiler/dialects/accfg.py:446]: good question! I originally wanted to have the lowering of the reset op in this PR as well, will remove!

File: compiler/transforms/accfg_insert_resets.py (hunk 3 above)
Comment (AntonLydike) [reply compiler/transforms/accfg_insert_resets.py:97]: Yes, correct, this docstring seems to be wrong. I do need to figure out exactly when the resets are inserted!

File: compiler/transforms/accfg_insert_resets.py (hunk 4 above)
Comment (AntonLydike) [reply compiler/transforms/accfg_insert_resets.py:105]: Me neither, will look at this tomorrow morning or so again when I'm fresh in the brain I think. So far it *seems* to work, according to the tests, but I'm not convinced the check is actually correct...

File: compiler/transforms/accfg_insert_resets.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    PatternRewriteWalker,
+    RewritePattern,
+)
+from xdsl.rewriter import InsertPoint
+from xdsl.traits import IsTerminator
+
+from compiler.dialects import accfg
+
+
+def ssa_val_rewrite_pattern(val_type: type[Attribute]):
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply compiler/transforms/accfg_insert_resets.py]: Done!

File: compiler/transforms/accfg_insert_resets.py (hunk 4 above)
Comment (AntonLydike) [reply compiler/transforms/accfg_insert_resets.py]: I cleaned it up, I have a much better feeling about this now than I had about the other code. Added more tests to reassure myself again.

File: compiler/accelerators/accelerator.py [hunk 7]
```diff
@@ -74,10 +74,3 @@ def lower_acc_setup(
         accelerator configuration.
         """
         raise NotImplementedError()
-
-    @staticmethod
-    @abstractmethod
-    def lower_reset(
-        rest_op: accfg.ResetOp, acc_op: accfg.AcceleratorOp
-    ) -> Sequence[Operation]:
```
Comment (JosseVanDelm) [inline compiler/accelerators/accelerator.py]: This is moving to a next PR, right?

File: tests/filecheck/transforms/accfg-insert-resets.mlir (hunk 5 above)
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:49]: I mainly thought it was weird that you already reset the accelerator's registers before the await.
I think we should be careful with this, since there could (and should) be hardware support for clearing all the registers (also the shadow registers) upon reset. That would make this code invalid if compiled. We should be very clear on semantics here, is all that I'm saying.
it is valid if the shadow registers are kept, but might lead to weird results if the reset affects them or other accelerator parameters. Maybe `reset` is too broad as a term?

File: compiler/transforms/accfg_insert_resets.py [hunk 8]
```diff
@@ -1,35 +1,49 @@
...
+    Callable[[_RewritePatternT, Operation, PatternRewriter], None],
+]:
+    """
+    Expresses a rewrite pattern that acts on SSA Values instead of operations.
+
+    Pass `val_type` and only match on SSA Values of the specified type.
+    """
+
+    def wrapper(
+        old_match_and_rewrite: Callable[
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_insert_resets.py:31]:
```suggestion
        wrapped_match_and_rewrite: Callable[
```

File: compiler/transforms/accfg_insert_resets.py [hunk 9]
```diff
@@ -99,12 +113,10 @@ class InsertResetsForDanglingStatesPattern(RewritePattern):
...
-        # abort if:
-        #  - there are uses of the op
-        #  - and all of them are setups/reset ops
-        # this may not be correct, let's see...
-        if val.uses and all(
-            isinstance(use.operation, accfg.SetupOp | accfg.ResetOp) for use in val.uses
+        # abort if
+        # any use is a reset or another setup op
+        if any(
+            isinstance(use.operation, accfg.ResetOp | accfg.SetupOp) for use in val.uses
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_insert_resets.py:119]: this makes way more sense! Thanks!

File: compiler/accelerators/accelerator.py (hunk 7 above)
Comment (AntonLydike) [reply compiler/accelerators/accelerator.py]: yes!

File: tests/filecheck/transforms/accfg-insert-resets.mlir (hunk 5 above)
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/transforms/accfg-insert-resets.mlir]: Added a flag `reset-after-await` to pass, will need some more fine tuning, but already most of the way there I think

File: compiler/transforms/accfg_insert_resets.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    # if no uses are found, just return the first position where the value is live:
+    if not val.uses:
+        if isinstance(val.owner, Operation):
+            yield InsertPoint.after(val.owner)
+        else:
+            yield InsertPoint.at_start(val.owner)
+        return
+
+    # remember insertion candidates and their positions
+    # we need to keep track of the last use of the value in each block it's used in.
```
[the remark this answers is not in the document]
Comment (AntonLydike) [reply compiler/transforms/accfg_insert_resets.py]: So that we know where it's dangling, basically.

File: compiler/inference/dataflow.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+def get_insertion_points_where_val_dangles(
+    vals: Sequence[SSAValue],
+) -> Iterable[InsertPoint]:
+    """
+    Return all insertion points for positions where a value starts dangling.
+
+    If val is used by a terminator, it's not assumed dangling in that block.
+
+    If multiple vals are given, return the earliest point in each block where
+    all vals used in that block are dangling.
```
Comment (JosseVanDelm) [inline compiler/inference/dataflow.py:60]: Why would we need to have multiple vals here? Is this because we both wait for the token and the state?
~~I find this quite difficult to read, would it make more sense to not make this a generator instead?~~
I should just read more generators haha

File: tests/filecheck/transforms/accfg-insert-resets.mlir [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        %0 = "accfg.launch"(%state) <{"param_names" = [], "accelerator" = "acc"}> : (!accfg.state<"acc">) -> !accfg.token<"acc">
+        "accfg.await"(%0) : (!accfg.token<"acc">) -> ()
+
+        yield
+    }, {
+        %1 = "accfg.launch"(%state) <{"param_names" = [], "accelerator" = "acc"}> : (!accfg.state<"acc">) -> !accfg.token<"acc">
+        "accfg.await"(%1) : (!accfg.token<"acc">) -> ()
+
+        yield
+    }) : (i1) -> ()
```
Comment (JosseVanDelm) [inline tests/filecheck/transforms/accfg-insert-resets.mlir:69]: Does it also work if the accelerator is not launched and awaited in one of the states?

File: tests/filecheck/transforms/accfg-insert-resets.mlir
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:1]: Is it possible you haven't added this yet?

File: compiler/transforms/accfg_insert_resets.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+        # if reset_after_await is given, reset after the tokens of the launch ops are no longer dangling
+        # (i.e. have been awaited)
+        if self.reset_after_await:
+            vals = []
+            for use in uses:
+                if isinstance(use.operation, accfg.LaunchOp):
+                    vals.append(use.operation.token)
+            if not vals:
+                vals.append(val)
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_insert_resets.py:87]: what does this do?

File: compiler/inference/dataflow.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        - yield %x
+        - arith.add %a, ...
+    """
+    for use in val.uses:
+        # return each use
+        yield use
+        # if it's an scf.for, recurse on block arg
+        if isinstance(use.operation, scf.For):
+            for_op = use.operation
+            yield from uses_through_controlflow(for_op.body.block.args[use.index - 2])
```
Comment (JosseVanDelm) [inline compiler/inference/dataflow.py:43]: I don't understand why we need the -2?
Don't we want something like this?
```suggestion
            yield from uses_through_controlflow(for_op.body.block.args[use.index - 2:])
```

File: compiler/transforms/accfg_insert_resets.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        ):
+            return
+
+        # if reset_after_await is given, reset after the tokens of the launch ops are no longer dangling
+        # (i.e. have been awaited)
+        if self.reset_after_await:
+            vals = []
+            for use in uses:
+                if isinstance(use.operation, accfg.LaunchOp):
+                    vals.append(use.operation.token)
```
Comment (JosseVanDelm) [inline compiler/transforms/accfg_insert_resets.py:85]: I think the setup state itself should also be appended here, no?

**Review by JosseVanDelm — APPROVED** [review]
I find this quite difficult to review :( so feel free if my comments don't make sense
I have a few more comments, the results look good though!

File: tests/filecheck/transforms/accfg-insert-resets.mlir
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:1]: Oh damn, you are right!

File: tests/filecheck/transforms/accfg-insert-resets.mlir (hunk 12 above)
Comment (AntonLydike) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:69]: Interesting question, will investigate tomorrow morning

File: compiler/inference/dataflow.py (hunk 14 above)
Comment (AntonLydike) [reply compiler/inference/dataflow.py:43]: Good point, yes. I should add more comments here.

The gist is that we want to get the block arg which represents the input arg, and -2 does that mapping. (`index - 3` is the idx of the loop carried variables, add one back because the first block arg is the loop index, and we end up at `index - 2`)

File: compiler/transforms/accfg_insert_resets.py (hunk 15 above)
Comment (AntonLydike) [reply compiler/transforms/accfg_insert_resets.py:85]: Hmmmmmm.

it doesn't make sense in my mind. But making that change breaks no tests, and I can't think of a case where it would break.

I'll give it some thought over lunch

File: tests/filecheck/transforms/accfg-insert-resets.mlir
[the remark this answers is not in the document]
Comment (AntonLydike) [reply tests/filecheck/transforms/accfg-insert-resets.mlir:1]: Added!

File: compiler/inference/dataflow.py (hunk 14 above)
Comment (JosseVanDelm) [reply compiler/inference/dataflow.py:47]: But so now it only works if there is just one block arg, right?

File: compiler/transforms/accfg_insert_resets.py (hunk 15 above)
Comment (JosseVanDelm) [reply compiler/transforms/accfg_insert_resets.py:85]: I think it might break this test that I'm proposing:
https://github.com/KULeuven-MICAS/snax-mlir/pull/173#discussion_r1681233566

Comment (JosseVanDelm) [reply compiler/transforms/accfg_insert_resets.py:85]: ping @AntonLydike, can you take a look at this?

**Comment by JosseVanDelm** [comment]
ping @AntonLydike !

File: compiler/inference/dataflow.py (hunk 14 above)
Comment (AntonLydike) [reply compiler/inference/dataflow.py]: I improved/fixed a bunch of stuff, feel free to have a look again!
